server/src/main/java/org/elasticsearch/index/query/TypeQueryV7Builder.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ParseField;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.core.RestApiVersion;
+import org.elasticsearch.index.mapper.MapperService;
+
+import java.io.IOException;
+
+public class TypeQueryV7Builder extends AbstractQueryBuilder<TypeQueryV7Builder> {
+    public static final String NAME = "type";
+    public static final ParseField NAME_V7 = new ParseField("type").forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7));
+    private static final ParseField VALUE_FIELD = new ParseField("value");
+    private static final ObjectParser<TypeQueryV7Builder, Void> PARSER = new ObjectParser<>(NAME, TypeQueryV7Builder::new);
+
+    static {
+        PARSER.declareString(QueryBuilder::queryName,
+            AbstractQueryBuilder.NAME_FIELD.forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)));
+        PARSER.declareFloat(QueryBuilder::boost,
+            AbstractQueryBuilder.BOOST_FIELD.forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)));
+        PARSER.declareString((queryBuilder, value) -> {},
+            VALUE_FIELD.forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)));
+    }
+
+    public TypeQueryV7Builder() {
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public TypeQueryV7Builder(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(VALUE_FIELD.getPreferredName(), MapperService.SINGLE_MAPPING_NAME);
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        return new MatchAllDocsQuery();
```

Comment:
We do actually still have type information available: `context.mappingLookup().getMapping().getRoot().name()`.  So you could check against the top-level type and then return a MatchAll/MatchNone depending on whether or not you get a match.

Reply:
so you suggest to check if the value used on a type query is _doc (this is the only type we can have, right?) then return matchAll otherwise matchNone?

Reply:
what values can I expect from `context.mappingLookup().getMapping().getRoot().name()` ?

Reply:
You can still specify non-`_doc` type names in 7x indexes.  `context .... name()` will return the type name of the current index.